File: java-lombok/src/test/java/it/gabrieletondi/telldontaskkata/useCase/OrderShipmentUseCaseTest.java
package it.gabrieletondi.telldontaskkata.useCase;

import it.gabrieletondi.telldontaskkata.domain.Order;
import it.gabrieletondi.telldontaskkata.domain.OrderStatus;
import it.gabrieletondi.telldontaskkata.doubles.TestOrderRepository;
import it.gabrieletondi.telldontaskkata.doubles.TestShipmentService;
import org.junit.Test;

import static org.hamcrest.Matchers.is;
import static org.hamcrest.Matchers.nullValue;
import static org.junit.Assert.assertThat;
import static org.junit.Assert.assertTrue;

public class OrderShipmentUseCaseTest {
    private final TestOrderRepository orderRepository = new TestOrderRepository();
    private final TestShipmentService shipmentService = new TestShipmentService();
    private final OrderShipmentUseCase useCase = new OrderShipmentUseCase(orderRepository, shipmentService);

    @Test
    public void shipsApprovedOrder() {
        Order initialOrder = createOrder(OrderStatus.APPROVED);
        orderRepository.addOrder(initialOrder);

        OrderShipmentRequest request = new OrderShipmentRequest();
        request.setOrderId(1);
        useCase.run(request);

        assertTrue(orderRepository.getSavedOrder().isShipped());
        assertThat(shipmentService.getShippedOrder(), is(initialOrder));
    }

    @Test(expected = OrderNotShippable.class)
    public void createdOrdersCannotBeShipped() {
        Order initialOrder = createOrder(OrderStatus.CREATED);
        orderRepository.addOrder(initialOrder);

        OrderShipmentRequest request = new OrderShipmentRequest();
        request.setOrderId(1);
        useCase.run(request);

        assertThat(orderRepository.getSavedOrder(), is(nullValue()));
        assertThat(shipmentService.getShippedOrder(), is(nullValue()));
    }

    @Test(expected = OrderNotShippable.class)
    public void rejectedOrdersCannotBeShipped() {
        Order initialOrder = createOrder(OrderStatus.REJECTED);
        orderRepository.addOrder(initialOrder);

        OrderShipmentRequest request = new OrderShipmentRequest();
        request.setOrderId(1);
        useCase.run(request);

        assertThat(orderRepository.getSavedOrder(), is(nullValue()));
        assertThat(shipmentService.getShippedOrder(), is(nullValue()));
    }

    @Test(expected = OrderAlreadyShipped.class)
    public void shippedOrdersCannotBeShippedAgain() {
        Order initialOrder = createOrder(OrderStatus.SHIPPED);
        orderRepository.addOrder(initialOrder);

        OrderShipmentRequest request = new OrderShipmentRequest();
        request.setOrderId(1);
        useCase.run(request);

        assertThat(orderRepository.getSavedOrder(), is(nullValue()));
        assertThat(shipmentService.getShippedOrder(), is(nullValue()));
    }

    private Order createOrder(OrderStatus status) {
        return Order.create(1, status);
    }
}

File: typescript/test/useCase/OrderShipmentUseCase.test.ts
import Order from "../../src/domain/Order";
import OrderShipmentRequest from "../../src/useCase/OrderShipmentRequest";
import TestOrderRepository from "../doubles/TestOrderRepository";
import OrderShipmentUseCase from "../../src/useCase/OrderShipmentUseCase";
import OrderCannotBeShippedError from "../../src/useCase/OrderCannotBeShippedError";
import OrderCannotBeShippedTwiceError from "../../src/useCase/OrderCannotBeShippedTwiceError";
import OrderStatus from "../../src/domain/OrderStatus";
import TestShipmentService from "../doubles/TestShipmentService";

describe('OrderShipmentUseCase should', () => {
    let orderRepository: TestOrderRepository;
    let shipmentService: TestShipmentService;
    let useCase: OrderShipmentUseCase;

    beforeEach(() => {
        orderRepository = new TestOrderRepository();
        shipmentService = new TestShipmentService();
        useCase = new OrderShipmentUseCase(orderRepository, shipmentService);
    });

    test('ship approved order', () => {
        let initialOrder = new Order(OrderStatus.APPROVED, 1);
        orderRepository.addOrder(initialOrder);

        useCase.run(new OrderShipmentRequest(1));

        expect(orderRepository.getSavedOrder()?.status).toBe(OrderStatus.SHIPPED);
        expect(shipmentService.getShippedOrder()).toBe(initialOrder);
    });

    test('not ship created orders', () => {
        orderRepository.addOrder(new Order(OrderStatus.CREATED, 1));

        expect(() => {
            useCase.run(new OrderShipmentRequest(1))
        }).toThrowError(OrderCannotBeShippedError);
        expect(orderRepository.getSavedOrder()).toBeUndefined();
        expect(shipmentService.getShippedOrder()).toBeUndefined();
    });

    test('not ship rejected orders', () => {
        orderRepository.addOrder(new Order(OrderStatus.REJECTED, 1));

        let request = new OrderShipmentRequest(1);

        expect(() => {
            useCase.run(request)
        }).toThrowError(OrderCannotBeShippedError);
        expect(orderRepository.getSavedOrder()).toBeUndefined();
        expect(shipmentService.getShippedOrder()).toBeUndefined();
    });

    test('not ship shipped orders', () => {
        orderRepository.addOrder(new Order(OrderStatus.SHIPPED, 1));

        expect(() => {
            useCase.run(new OrderShipmentRequest(1))
        }).toThrowError(OrderCannotBeShippedTwiceError);
        expect(orderRepository.getSavedOrder()).toBeUndefined();
        expect(shipmentService.getShippedOrder()).toBeUndefined();
    });
});
